### ISSUE TITLE
Optimizing for Waterfall Displays?

### ISSUE THREAD

**Issue by morta5** [mastodon/mastodon-android#1]
I just opened the app for the first time(in fullscreen) on my Moto Edge Plus and noticed it's not optimized for Waterfall Displays. I can use it normally after disabling the edge screen but it would of course be better if optimized. 
See the attached screenshots
![Screenshot_20220413-194748](https://user-images.githubusercontent.com/66527936/163240713-026579cd-7d7d-415c-8f73-f08e31abfc01.png)
![Screenshot_20220413-194733](https://user-images.githubusercontent.com/66527936/163240717-4ae03fca-06f1-46a5-89e9-a2736a658879.png)
![Screenshot_20220413-194727](https://user-images.githubusercontent.com/66527936/163240725-ace42164-be29-4b43-9baa-1facab6d418a.png)
![Screenshot_20220413-194730](https://user-images.githubusercontent.com/66527936/163240729-a1b4d41f-d070-4a67-bd76-2ff70a0c4a90.png)
![Screenshot_20220413-194704](https://user-images.githubusercontent.com/66527936/163240732-47c7480a-b351-474a-81a5-7733c837de60.png)
![Screenshot_20220413-194424-599](https://user-images.githubusercontent.com/66527936/163240733-760e32db-6d31-4a0d-83e7-521536472749.png)
![Screenshot_20220413-194419](https://user-images.githubusercontent.com/66527936/163240736-895d6ad8-6b87-45ed-a9d8-aad4c03ced16.png)

**Comment by grishka** [comment]
TIL that this device exists. It looks pretty similar to the Galaxy S9 I mostly debug the app on.

But, things getting beneath the status bar is something I'd like to avoid. Do they do something special with the window insets (it feels like they do)? Do they have any developer docs about that?

**Comment by morta5** [comment]
https://developer.android.com/reference/android/view/DisplayCutout.html#getWaterfallInsets()

**Comment by grishka** [comment]
That's about the edges, but doesn't explain why `systemWindowInsets` seemingly aren't returned correctly 🤔

**Comment by morta5** [comment]
Not sure but after tapping on the settings-icon and going back it is getting displayed correctly

**Comment by grishka** [comment]
Okay. Does it also fix itself if you rotate the screen? And it's Android 11 and stock ROM, right?

It might well be that my handling of window insets isn't 100% correct, so for now I'll wait to see if this issue happens on any other devices. In particular, I only have one device with a screen cutout (Pixel 4a), so my testing in this area is somewhat limited.

**Comment by morta5** [comment]
No, it doesn't fix itself when rotating. 
Btw can't you just emulate the waterfall in developer settings?
![Screenshot_20220413-211748](https://user-images.githubusercontent.com/66527936/163254138-f98a346b-43eb-4713-9e86-4d34a41f5371.png)

**Comment by morta5** [comment]
It does only happen when using the "fullscreen" option to open the app, which should enable(or force idk.) the waterfall area.
![Screenshot_20220413-212928-327](https://user-images.githubusercontent.com/66527936/163256763-9c0f9250-c3d5-4430-a5ae-29c5df1ed491.png)

**Comment by grishka** [comment]
I'm leaning towards this being something Motorola did that can't be reproduced on AOSP/Pixel, but as I said, I'd like to collect more reports to see how many, and which, devices are affected, if any.

> It does only happen when using the "fullscreen" option to open the app, which should enable(or force idk.) the waterfall area.

It does look like an OEM addition. There's no such thing on the emulator.

![screenshot](https://user-images.githubusercontent.com/1478258/163257214-fa512f24-ec3c-4fdc-aab8-497af76c3b3b.png)

**Comment by morta5** [comment]
Sure, but do you also have (black)/#282c37/unused bars at the right and left of the screen on your S9 and are these intended? 
The toolbar thing wasn't actually why I opened this issue, it was about the unused waterfall area(which is probably why this sometimes happens).

**Comment by grishka** [comment]
> Sure, but do you also have (black)/#282c37/unused bars at the right and left of the screen on your S9 and are these intended?

Nope. But then it runs Android 9, and waterfall insets were only added in 10.

**Comment by grishka** [comment]
Anyway, I was bored and I wanted something similar for quite some time, so I made a simple app that visualizes the window insets in an easy to understand form. Could you please install it and take screenshots in both modes?

<details>
  <summary>For those curious, here it is running on different devices</summary>

![screenshot](https://user-images.githubusercontent.com/1478258/163282024-d179a337-e8ca-47b7-a92b-5d01e230b384.png)

![screenshot](https://user-images.githubusercontent.com/1478258/163282081-4d98e5be-ac1b-4548-a94b-0b460313ab45.png)

![screenshot](https://user-images.githubusercontent.com/1478258/163282150-1a497d79-3f3a-46ba-9b3a-c78cee2028c0.png)

![screenshot](https://user-images.githubusercontent.com/1478258/163282176-47444c01-307b-4e81-a5e3-c3c902013b69.png)

</details>

[app-debug.apk.zip](https://github.com/mastodon/mastodon-android/files/8485055/app-debug.apk.zip) (remove .zip after downloading because github is picky about file types)

**Comment by morta5** [comment]
Sure. Here are all screenshots in both modes (but there doesn't seem to be any difference between fullscreen and normal):
![Screenshot_20220414-011002](https://user-images.githubusercontent.com/66527936/163284582-63152a6e-f081-4bc9-a8eb-6af422ba4efe.png)
![Screenshot_20220414-010955](https://user-images.githubusercontent.com/66527936/163284585-d0731c60-5579-4613-8310-64247f60edb5.png)
![Screenshot_20220414-010937](https://user-images.githubusercontent.com/66527936/163284587-1f1c2e79-fc4f-4abe-97fa-d2766139fbd3.png)
![Screenshot_20220414-010919](https://user-images.githubusercontent.com/66527936/163284588-ec22390a-288f-4fe8-9c62-5685e15b29ed.png)
![Screenshot_20220414-010906](https://user-images.githubusercontent.com/66527936/163284589-4d16ad99-bc1f-4fe3-be9e-23488d559b9e.png)
![Screenshot_20220414-010850](https://user-images.githubusercontent.com/66527936/163284592-ea06ff92-dbb2-46cd-88f2-d496c9d4e402.png)
![Screenshot_20220414-010813](https://user-images.githubusercontent.com/66527936/163284594-3ea7ad7c-002a-43b0-9656-936e2262551d.png)
![Screenshot_20220414-010732](https://user-images.githubusercontent.com/66527936/163284595-31ab69e3-10c9-4b83-b45c-b2f318197ae4.png)

**Comment by grishka** [comment]
Thanks!

Setting `LAYOUT_IN_DISPLAY_CUTOUT_MODE_ALWAYS` was probably not a good idea (and the Mastodon app doesn't do it). Try this one, there should be a difference in those modes with the checkbox unchecked:
[app-debug.apk.zip](https://github.com/mastodon/mastodon-android/files/8485158/app-debug.apk.zip)

**Comment by morta5** [comment]
Those marked with "F" were made in fullscreen mode
![Screenshot_20220414-013154-272](https://user-images.githubusercontent.com/66527936/163286597-c1c82a90-3594-4fe0-9eb2-6f7aea15eb43.png)
![Screenshot_20220414-013139-551](https://user-images.githubusercontent.com/66527936/163286598-8568f237-622f-465b-8cb9-fbb8515b69eb.png)
![Screenshot_20220414-013110](https://user-images.githubusercontent.com/66527936/163286601-777c39ad-09a7-42b6-b0a5-685c2c800592.png)
![Screenshot_20220414-013104](https://user-images.githubusercontent.com/66527936/163286603-bd575f8d-28d7-4bc5-ab62-49a48b419622.png)
![Screenshot_20220414-013049](https://user-images.githubusercontent.com/66527936/163286604-8de33441-803a-4efd-b1e0-e1ea48c9e7ac.png)
![Screenshot_20220414-013041](https://user-images.githubusercontent.com/66527936/163286606-469399bd-4c9d-4b86-afdf-d7674183c485.png)
![Screenshot_20220414-013030-241](https://user-images.githubusercontent.com/66527936/163286607-cc6eff18-dc68-4094-b845-c70087bcad21.png)
![Screenshot_20220414-013016-580](https://user-images.githubusercontent.com/66527936/163286610-a56863cd-3f50-4c16-b555-6f48a5295c5e.png)

**Comment by grishka** [comment]
Hm. Yes it does look like I expect it to (the top system inset is there), the toolbar getting behind the status bar must be a bug in the way I handle insets then. Also strange how no DisplayCutout is returned at all when you don't set that layout mode 🤔

**Comment by morta5** [comment]
> Also strange how no DisplayCutout is returned at all when you don't set that layout mode 🤔

That seems to be a bug. I tried it a second time and now I am seeing the display cutout.
![Screenshot_20220414-015013~2](https://user-images.githubusercontent.com/66527936/163288040-c89cfc5a-8e71-40a6-931a-7d92ea4e9fae.png)
![Screenshot_20220414-015055](https://user-images.githubusercontent.com/66527936/163288044-8486c0ab-68f5-4b55-8f14-0577361e4526.png)
